File: apps/agent/src/orchestrator.ts
export type CallContextRequest = {
  callId: string;
  phoneNumber: string;
  direction: "inbound";
};

export type CustomerCallContext = {
  nameCustomer: string;
};

export type TransferTargetRequest = {
  callId: string;
  phoneNumber: string;
  reason: string;
};

export type TransferTarget =
  | { available: false }
  | {
      available: true;
      transferTo: string;
      agentName: string;
    };

export type CallOrchestratorClient = {
  getCallContext(
    request: CallContextRequest,
  ): Promise<CustomerCallContext>;
  getTransferTarget(request: TransferTargetRequest): Promise<TransferTarget>;
};

type FetchImplementation = (
  input: string,
  init?: RequestInit,
) => Promise<Response>;

export function createCallOrchestratorClient(
  baseUrl: string,
  fetchImplementation: FetchImplementation = fetch,
): CallOrchestratorClient {
  const normalizedBaseUrl = baseUrl.replace(/\/+$/, "");

  return {
    async getCallContext(request) {
      const response = await fetchImplementation(
        `${normalizedBaseUrl}/api/call-context`,
        {
        method: "POST",
        headers: { "content-type": "application/json" },
        body: JSON.stringify(request),
        signal: AbortSignal.timeout(2_000),
        },
      );

      if (!response.ok) {
        throw new Error(
          `Call orchestrator returned HTTP ${response.status}`,
        );
      }

      const body = (await response.json()) as Partial<CustomerCallContext>;
      const nameCustomer = body.nameCustomer?.trim();
      if (!nameCustomer) {
        throw new Error("Call orchestrator response is missing nameCustomer");
      }

      return { nameCustomer };
    },

    async getTransferTarget(request) {
      const response = await fetchImplementation(
        `${normalizedBaseUrl}/api/transfer-target`,
        {
          method: "POST",
          headers: { "content-type": "application/json" },
          body: JSON.stringify(request),
          signal: AbortSignal.timeout(2_000),
        },
      );

      if (!response.ok) {
        throw new Error(
          `Call orchestrator returned HTTP ${response.status}`,
        );
      }

      const body = (await response.json()) as Partial<TransferTarget>;
      if (body.available === false) {
        return { available: false };
      }

      const transferTo =
        "transferTo" in body ? body.transferTo?.trim() : undefined;
      const agentName =
        "agentName" in body ? body.agentName?.trim() : undefined;
      if (body.available !== true || !transferTo || !agentName) {
        throw new Error("Call orchestrator returned an invalid transfer target");
      }

      return { available: true, transferTo, agentName };
    },
  };
}

export type SipParticipantInfo = {
  identity: string;
  attributes: Record<string, string>;
};

export async function prepareInboundCallContext(
  participant: SipParticipantInfo,
  client: Pick<CallOrchestratorClient, "getCallContext">,
): Promise<CustomerCallContext> {
  const callId =
    participant.attributes["sip.callID"] || `call-${participant.identity}`;
  const phoneNumber =
    participant.attributes["sip.phoneNumber"] || participant.identity;

  try {
    return await client.getCallContext({
      callId,
      phoneNumber,
      direction: "inbound",
    });
  } catch (error) {
    console.warn(
      "Call orchestrator lookup failed; using fallback customer context",
      error instanceof Error ? error.message : error,
    );
    return { nameCustomer: "Quý khách" };
  }
}

File: apps/agent-py/orchestrator.py
import asyncio
import json
import os
import re
import urllib.error
import urllib.request
from dataclasses import dataclass
from typing import Mapping


FALLBACK_CUSTOMER_NAME = "Quý khách"
DEFAULT_CALL_ORCHESTRATOR_URL = "http://call-orchestrator:3002"


@dataclass(frozen=True)
class CustomerCallContext:
    name_customer: str


def normalize_prompt_value(value: str) -> str:
    return re.sub(r"[\r\n\t]+", " ", value).strip()[:120]


def get_call_orchestrator_url() -> str:
    return os.environ.get("CALL_ORCHESTRATOR_URL", "").strip() or DEFAULT_CALL_ORCHESTRATOR_URL


def _post_call_context(endpoint: str, payload: dict[str, str]) -> CustomerCallContext:
    body = json.dumps(payload).encode("utf-8")
    request = urllib.request.Request(
        endpoint,
        data=body,
        headers={"content-type": "application/json"},
        method="POST",
    )

    with urllib.request.urlopen(request, timeout=2) as response:
        if response.status < 200 or response.status >= 300:
            raise RuntimeError(f"Call orchestrator returned HTTP {response.status}")

        data = json.loads(response.read().decode("utf-8"))

    name_customer = normalize_prompt_value(str(data.get("nameCustomer", "")))
    if not name_customer:
        raise RuntimeError("Call orchestrator response is missing nameCustomer")

    return CustomerCallContext(name_customer=name_customer)


async def get_call_context(call_id: str, phone_number: str) -> CustomerCallContext:
    base_url = get_call_orchestrator_url().rstrip("/")
    endpoint = f"{base_url}/api/call-context"
    payload = {
        "callId": call_id,
        "phoneNumber": phone_number,
        "direction": "inbound",
    }

    return await asyncio.to_thread(_post_call_context, endpoint, payload)


def participant_attributes(participant: object) -> Mapping[str, str]:
    attributes = getattr(participant, "attributes", None)
    if isinstance(attributes, Mapping):
        return attributes
    return {}


async def prepare_inbound_call_context(participant: object) -> CustomerCallContext:
    identity = str(getattr(participant, "identity", "") or "")
    attributes = participant_attributes(participant)
    call_id = attributes.get("sip.callID") or f"call-{identity}"
    phone_number = attributes.get("sip.phoneNumber") or identity

    try:
        return await get_call_context(call_id, phone_number)
    except (OSError, urllib.error.URLError, TimeoutError, RuntimeError, ValueError, json.JSONDecodeError) as error:
        print(f"Call orchestrator lookup failed; using fallback customer context: {error}")
        return CustomerCallContext(name_customer=FALLBACK_CUSTOMER_NAME)
